File: routing-v4/src/llm_client.py
"""
llm_client.py  —  routing-v4
Failover-aware LiteLLM async client.

After every call the result is reported back to the FailoverManager so
circuit-breaker state is updated in real time.
"""

from __future__ import annotations
import asyncio
import logging
import os
import time

import litellm
from litellm import Router

from src.failover_manager import failover_manager

litellm.success_callback = []
litellm.set_verbose = False
logging.getLogger("LiteLLM").setLevel(logging.ERROR)

OLLAMA_BASE_URL = os.getenv("OLLAMA_BASE_URL", "http://localhost:11434")
OLLAMA_TIMEOUT  = int(os.getenv("OLLAMA_TIMEOUT", "120"))

_TIER_MODELS: dict[str, str] = {
    "model::light":    os.getenv("LLM_LIGHT",    "granite4.1:3b"),
    "model::medium":   os.getenv("LLM_MEDIUM",   "llama3.2:latest"),
    "model::balanced": os.getenv("LLM_BALANCED", "gemma3:4b"),
    "model::heavy":    os.getenv("LLM_HEAVY",    "mistral-small3.2:latest"),
}

# Fallback model names (secondary deployments for LiteLLM Router)
_FALLBACK_MODELS: dict[str, str] = {
    "model::light":    os.getenv("LLM_LIGHT_FALLBACK",    "llama3.2:latest"),
    "model::medium":   os.getenv("LLM_MEDIUM_FALLBACK",   "gemma3:4b"),
    "model::balanced": os.getenv("LLM_BALANCED_FALLBACK", "llama3.2:latest"),
    "model::heavy":    os.getenv("LLM_HEAVY_FALLBACK",    "gemma3:4b"),
}


def _build_model_list() -> list[dict]:
    """
    Build LiteLLM Router model list with 2 deployments per tier label:
      - Primary deployment (the configured Ollama model)
      - Fallback deployment (secondary, different Ollama model)
    This gives LiteLLM its own layer of retry/load-balance within a tier,
    while our FailoverManager handles cross-tier failover.
    """
    entries: list[dict] = []
    for tier_label, primary_model in _TIER_MODELS.items():
        entries.append({
            "model_name": tier_label,
            "litellm_params": {
                "model":    f"ollama/{primary_model}",
                "api_base": OLLAMA_BASE_URL,
                "timeout":  OLLAMA_TIMEOUT,
            },
        })
        fallback_model = _FALLBACK_MODELS.get(tier_label)
        if fallback_model and fallback_model != primary_model:
            entries.append({
                "model_name": tier_label,
                "litellm_params": {
                    "model":    f"ollama/{fallback_model}",
                    "api_base": OLLAMA_BASE_URL,
                    "timeout":  OLLAMA_TIMEOUT,
                },
            })
    return entries


_router = Router(
    model_list=_build_model_list(),
    routing_strategy="latency-based-routing",
    num_retries=2,
    retry_after=2,
    allowed_fails=3,
    cooldown_time=30,
    set_verbose=False,
)


def get_router() -> Router:
    return _router


async def call_model(
    model_label: str,
    prompt: str,
    max_tokens: int = 1024,
) -> tuple[str, float, int]:
    """
    Call a model tier via LiteLLM Router.
    Reports success/failure to the FailoverManager for circuit-breaker tracking.

    Returns (response_text, latency_seconds, output_tokens).
    """
    messages = [{"role": "user", "content": prompt}]
    t0 = time.perf_counter()
    try:
        response = await _router.acompletion(
            model=model_label,
            messages=messages,
            max_tokens=max_tokens,
            temperature=0.3,
        )
        latency = time.perf_counter() - t0
        text    = response.choices[0].message.content or ""
        tokens  = (
            response.usage.completion_tokens
            if response.usage
            else max(1, len(text) // 4)
        )
        failover_manager.record_success(model_label)
        return text, latency, tokens

    except Exception as exc:
        latency = time.perf_counter() - t0
        failover_manager.record_failure(model_label)
        real_model = _TIER_MODELS.get(model_label, model_label)
        return (
            f"[LiteLLM ERROR — {real_model}: {exc}]",
            latency,
            0,
        )


def call_model_sync(
    model_label: str,
    prompt: str,
    max_tokens: int = 1024,
) -> tuple[str, float, int]:
    return asyncio.run(call_model(model_label, prompt, max_tokens))


TIER_TO_MODEL: dict[str, str] = _TIER_MODELS

File: v5-go-implementation/internal/llmclient/llmclient.go
// Package llmclient provides a failover-aware HTTP client for Ollama via the
// OpenAI-compatible chat completions API.
package llmclient

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"os"
	"time"

	"github.com/routing-v5/internal/failover"
)

// env-configurable
var (
	ollamaBase = func() string {
		if v := os.Getenv("OLLAMA_BASE_URL"); v != "" {
			return v
		}
		return "http://localhost:11434"
	}()

	ollamaTimeout = func() time.Duration {
		if v := os.Getenv("OLLAMA_TIMEOUT"); v != "" {
			var n int
			fmt.Sscanf(v, "%d", &n)
			return time.Duration(n) * time.Second
		}
		return 120 * time.Second
	}()
)

// TierModels maps the tier label to the primary Ollama model name.
var TierModels = map[string]string{
	"model::light":    getenv("LLM_LIGHT", "granite4.1:3b"),
	"model::medium":   getenv("LLM_MEDIUM", "llama3.2:latest"),
	"model::balanced": getenv("LLM_BALANCED", "gemma3:4b"),
	"model::heavy":    getenv("LLM_HEAVY", "mistral-small3.2:latest"),
}

// fallbackModels maps tier labels to secondary Ollama model names.
var fallbackModels = map[string]string{
	"model::light":    getenv("LLM_LIGHT_FALLBACK", "llama3.2:latest"),
	"model::medium":   getenv("LLM_MEDIUM_FALLBACK", "gemma3:4b"),
	"model::balanced": getenv("LLM_BALANCED_FALLBACK", "llama3.2:latest"),
	"model::heavy":    getenv("LLM_HEAVY_FALLBACK", "gemma3:4b"),
}

func getenv(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

// openAIRequest is the JSON body sent to the completions endpoint.
type openAIRequest struct {
	Model       string           `json:"model"`
	Messages    []openAIMessage  `json:"messages"`
	MaxTokens   int              `json:"max_tokens"`
	Temperature float64          `json:"temperature"`
	Stream      bool             `json:"stream"`
}

type openAIMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type openAIResponse struct {
	Choices []struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	} `json:"choices"`
	Usage struct {
		CompletionTokens int `json:"completion_tokens"`
	} `json:"usage"`
}

// CallResult holds the result of a single model call.
type CallResult struct {
	Text      string
	LatencySec float64
	Tokens    int
}

// CallModel calls the Ollama model identified by tierLabel.
// It reports success/failure to the FailoverManager and retries once on
// transient errors before returning an error string.
func CallModel(
	ctx context.Context,
	fm *failover.Manager,
	tierLabel string,
	prompt string,
	maxTokens int,
) CallResult {
	ollamaModel := TierModels[tierLabel]
	start := time.Now()

	result, err := doCall(ctx, ollamaModel, prompt, maxTokens)
	if err != nil {
		// One retry with fallback model
		fb := fallbackModels[tierLabel]
		if fb != "" && fb != ollamaModel {
			result, err = doCall(ctx, fb, prompt, maxTokens)
		}
	}

	latency := time.Since(start).Seconds()
	if err != nil {
		fm.RecordFailure(tierLabel)
		return CallResult{
			Text:       fmt.Sprintf("[LiteLLM ERROR — %s: %v]", ollamaModel, err),
			LatencySec: latency,
			Tokens:     0,
		}
	}
	fm.RecordSuccess(tierLabel)
	tokens := result.Usage.CompletionTokens
	text := ""
	if len(result.Choices) > 0 {
		text = result.Choices[0].Message.Content
	}
	if tokens == 0 {
		tokens = int(math.Max(1, float64(len(text)/4)))
	}
	return CallResult{Text: text, LatencySec: latency, Tokens: tokens}
}

func doCall(ctx context.Context, model, prompt string, maxTokens int) (*openAIResponse, error) {
	body := openAIRequest{
		Model:       model,
		Messages:    []openAIMessage{{Role: "user", Content: prompt}},
		MaxTokens:   maxTokens,
		Temperature: 0.3,
		Stream:      false,
	}
	data, err := json.Marshal(body)
	if err != nil {
		return nil, err
	}
	reqCtx, cancel := context.WithTimeout(ctx, ollamaTimeout)
	defer cancel()

	req, err := http.NewRequestWithContext(reqCtx, http.MethodPost,
		ollamaBase+"/v1/chat/completions", bytes.NewReader(data))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(b))
	}
	var result openAIResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return &result, nil
}
